### ISSUE TITLE
Utilize aws s3 ls instead of s3api head-bucket to check for access before barman-cloud-backup execution

### ISSUE THREAD

**Issue by benjamin-schilling-csq** [EnterpriseDB/barman#929]
barman-cloud-backup currently uses the head-bucket operation of the s3api to check for existence and access to the targeted S3 bucket. For those trying to write restricted policy to prefixes within the bucket, the only thing that allows head-bucket to properly work is to allow ListBucket to the entirety of the bucket. If using the equivalent of aws-cli s3 ls, the policy can allow for checking that the root of the bucket exists while also restricting prefixes levels below the root of the bucket, permitting for a more secure solution. 

If this is a feasible feature request it would be greatly appreciated if it could be implemented.

**Comment by mwijngaard** [comment]
I also just ran into this issue as I restrict s3:ListBucket to prefixes. Since the current implementation merely uses HeadBucket to check for bucket existence, the same could also be achieved by using ListObjectsV2 with a `prefix` parameter. This returns a 404 for non-existent buckets, but is compatible with prefix-restricted s3:ListBucket permissions. This seems like a simple method replace in `_check_bucket_existence` without any downside, or is there something I'm missing @martinmarques ?

**Comment by mwijngaard** [comment]
@andremagui I would probably opt against regarding AccessDenied as a 'bucket exists' and just require the bucket to be listable. Since this is also the current requirement (head-bucket requires ListBucket without prefix, which also always allows ListObjectsV2), that would be backwards compatible, and it just makes more sense to me. 

Also, it seems to me that an explicit 'deny' would not be necessary in any case, since deny is the default? Merely having an allow role with prefix condition would be sufficient?

**Comment by andremagui** [comment]
Hi @mwijngaard, we're currently discussing this internally and will get back to you by early next week at the latest.

**Comment by DimitriosLisenko** [comment]
Hi all! I've implemented this change because we need it internally - you can see the update in https://github.com/EnterpriseDB/barman/pull/1094. Happy to make any changes that might be necessary for this to be merged.

**Comment by martinmarques** [comment]
Has anyone looked at the replies to this issue from Terraform?

https://github.com/hashicorp/terraform-provider-aws/issues/17195

**Comment by mwijngaard** [comment]
@martinmarques which part exactly? The comment mentioning the null condition? Like the last comment says, that would indeed allow listing the full bucket, which is the opposite of what we're trying to achieve.

**Comment by mwijngaard** [comment]
@andremagui what @DimitriosLisenko has implemented is exactly what I had in mind.

**Comment by andremagui** [comment]
We pulled @DimitriosLisenko PR, tested it and added a few things needed. This will be in the next release. Thanks a lot everyone for the patch request and contribution @DimitriosLisenko @benjamin-schilling-csq @mwijngaard

**Comment by DimitriosLisenko** [comment]
@andremagui @martinmarques Hello! I can see that a new issue was opened [here](https://github.com/EnterpriseDB/barman/issues/1101) re: reducing expensive S3 operations, as a result of which the list_objects_v2 change was reverted.

We really need that functionality since we can't afford to let the root of our S3 bucket be listable - is there any chance of this getting re-added, or perhaps determined via a configuration item (i.e. use head_bucket or list_buckets_v2 depending on some configuration)? Or is there some other work that needs to be done first (which I'd be happy to assist with as well)?

**Comment by andremagui** [comment]
Hi @DimitriosLisenko, thanks for raising this. We’re currently discussing this internally and will evaluate the best approach for handling this case. Our goal is to find a solution that balances functionality with minimizing expensive S3 operations. We’re aiming to have a fix in place for the next release.

**Comment by alexklibisz** [comment]
Hi all, is it possible to add a command line argument or some other configuration to skip this connection check entirely? 

I'm particularly interested in optimizing the barman-cloud-wal-archive command as part of a cloudnative-pg deployment. In this case, I don't understand what value this connection check is adding for all the cost. If the connection check fails, what does barman do differently based on that result? It seems like barman just exists with a network error ([source](https://github.com/EnterpriseDB/barman/blob/9051dc0090df12052923e820451461a9f103eab3/barman/clients/cloud_check_wal_archive.py#L47-L49)). Why not just try the the object upload and see how it goes?
